core-plugins/src/main/java/co/cask/hydrator/plugin/batch/joiner/Joiner.java

```diff
@@ -0,0 +1,289 @@
+          if (Strings.isNullOrEmpty(entry.getKey()) || Strings.isNullOrEmpty(entry.getValue())) {
+            throw new IllegalArgumentException("Either field name or alias is missing.");
+          }
+          perStageFieldsToRename.put(stageToRenameField.getKey(), entry.getKey(), entry.getValue());
+        }
+      }
+    } else {
+      throw new IllegalArgumentException("selectedFields can not be empty. Please provide at least 1 selectedFields");
+    }
+    return perStageFieldsToRename;
+  }
+
+  Iterable<String> getRequiredInputs() {
+    if (conf.getRequiredInputs() != null) {
+      return Splitter.on(',').trimResults().omitEmptyStrings().split(conf.getRequiredInputs());
+    }
+    return Collections.<String>emptyList();
+  }
+
+  @Path("outputSchema")
+  public Schema getOutputSchema(GetSchemaRequest request) {
+    try {
+      return getOutputSchema(request.inputSchemas);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(e.getMessage());
+    }
+  }
+
+  /**
+   * Endpoint request for output schema.
+   */
+  public static class GetSchemaRequest extends JoinerConfig {
+    public Map<String, Schema> inputSchemas;
+  }
+
+
+  private Schema getOutputSchema(Map<String, Schema> inputSchemas) {
+    validateRequiredInputs(inputSchemas);
+    // sort the input schemas by input names to get the deterministic order of fields for output schema
+    Map<String, Schema> inputs = new TreeMap<>(inputSchemas);
+    Map<String, Schema.Field> outputFields = new LinkedHashMap<>();
+    Map<String, String> duplicateFieldsMap = new HashMap<>();
+
+    for (Map.Entry<String, Schema> input : inputs.entrySet()) {
+      String stageName = input.getKey();
+      Schema inputSchema = input.getValue();
+      Map<String, String> fieldsToSelectRename = perStageFieldsToRename.row(stageName);
+
+      for (Schema.Field inputField : inputSchema.getFields()) {
+        String inputFieldName = inputField.getName();
+        // drop the field if not part of fieldsToSelectRename
+        if (!fieldsToSelectRename.containsKey(inputFieldName)) {
+          continue;
+        }
+        // rename the out field
+        String outputFieldName = fieldsToSelectRename.get(inputFieldName);
+
+        // check if there are any duplicate fields
+        if (outputFields.containsKey(outputFieldName)) {
+          duplicateFieldsMap.put(stageName, com.google.common.base.Joiner.on(",")
+            .join(duplicateFieldsMap.get(stageName), outputFieldName));
+          continue;
+        }
+
+        // set nullable fields for non-required inputs
+        if (!Iterables.isEmpty(requiredInputs) && Iterables.contains(requiredInputs, stageName)) {
+          outputFields.put(outputFieldName, Schema.Field.of(outputFieldName, inputField.getSchema()));
+        } else {
+          outputFields.put(outputFieldName, Schema.Field.of(outputFieldName, Schema.nullableOf(inputField.getSchema()
+          )));
```

Comment:
just putting the close-parens on the next line seems strange. can you split the statement somewhere else?